src/Servers/HttpSys/src/AsyncAcceptContext.cs

```diff
@@ -2,121 +2,132 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
-using System.Diagnostics.CodeAnalysis;
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.AspNetCore.Http;
+using System.Threading.Tasks.Sources;
 using Microsoft.AspNetCore.HttpSys.Internal;
 
 namespace Microsoft.AspNetCore.Server.HttpSys
 {
-    internal unsafe class AsyncAcceptContext : TaskCompletionSource<RequestContext>, IDisposable
+    internal unsafe class AsyncAcceptContext : IValueTaskSource<RequestContext>, IDisposable
     {
-        internal static readonly IOCompletionCallback IOCallback = new IOCompletionCallback(IOWaitCallback);
+        private static readonly IOCompletionCallback IOCallback = IOWaitCallback;
+        private readonly PreAllocatedOverlapped _preallocatedOverlapped;
+
+        // mutable struct; do not make this readonly
+        private ManualResetValueTaskSourceCore<RequestContext> _tcs = new()
+        {
+            // We want to run continuations on the IO threads
+            RunContinuationsAsynchronously = false
+        };
 
         private NativeRequestContext _nativeRequestContext;
 
         internal AsyncAcceptContext(HttpSysListener server)
         {
             Server = server;
-            AllocateNativeRequest();
+            _preallocatedOverlapped = new(IOCallback, state: this, pinData: null);
         }
 
         internal HttpSysListener Server { get; }
 
-        [SuppressMessage("Microsoft.Design", "CA1031:DoNotCatchGeneralExceptionTypes", Justification = "Redirecting to callback")]
-        [SuppressMessage("Microsoft.Reliability", "CA2000:Dispose objects before losing scope", Justification = "Disposed by callback")]
+        internal ValueTask<RequestContext> AcceptAsync()
+        {
+            _tcs.Reset();
+
+            AllocateNativeRequest();
+
+            uint statusCode = QueueBeginGetContext();
+            if (statusCode != UnsafeNclNativeMethods.ErrorCodes.ERROR_SUCCESS &&
+                statusCode != UnsafeNclNativeMethods.ErrorCodes.ERROR_IO_PENDING)
+            {
+                // some other bad error, possible(?) return values are:
```

Comment:
No disposal is done here since the caller will end up calling AcceptAsync again and that will end up clearing the previously allocated request context (same goes for all other places that used to call dispose).